src/transform-sdk/rust/sr-sys/src/lib.rs

```diff
@@ -123,3 +130,135 @@ impl SchemaRegistryClientImpl for AbiSchemaRegistryClient {
         ))
     }
 }
+
+trait Cache {
+    type Key;
+    type Value;
+    fn get(&self, k: &Self::Key) -> Option<&Self::Value>;
+    fn put(&self, k: Self::Key, v: Self::Value) -> Option<Self::Value>;
+}
+
+#[derive(Debug)]
+struct LruCache<K, V> {
+    underlying: UnsafeCell<LruCacheImpl<K, V>>,
+}
+
+impl<K: HashT + Eq, V> LruCache<K, V> {
+    pub fn new(max_entries: Option<usize>) -> LruCache<K, V> {
+        let nz_max = match max_entries {
+            Some(v) if v > 0 => NonZeroUsize::new(v),
+            _ => NonZeroUsize::new(10),
+        };
+        Self {
+            underlying: UnsafeCell::new(LruCacheImpl::<K, V>::new(nz_max.unwrap())),
+        }
+    }
+}
+
+impl<K: HashT + Eq, V> Cache for LruCache<K, V> {
+    type Key = K;
+    type Value = V;
+
+    fn get(&self, k: &Self::Key) -> Option<&Self::Value> {
+        unsafe {
+            return (*self.underlying.get()).get(k);
+        }
+    }
+
+    fn put(&self, k: Self::Key, v: Self::Value) -> Option<Self::Value> {
+        unsafe {
+            return (*self.underlying.get()).put(k, v);
+        }
+    }
+}
+
+/// SchemaId is a type for the registered ID of a schema.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct SubjectVersion {
+    subject: String,
+    version: SchemaVersion,
+}
+
+impl SubjectVersion {
+    pub fn new(subject: impl Into<String>, version: SchemaVersion) -> Self {
+        Self {
+            subject: subject.into(),
+            version,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CachingSchemaRegistryClient {
+    underlying: AbiSchemaRegistryClient,
+    schema_by_id_cache: LruCache<SchemaId, Schema>,
+    schema_by_sv_cache: LruCache<SubjectVersion, SubjectSchema>,
+}
+
+impl CachingSchemaRegistryClient {
```

Comment:
Can we move all this into the `sr` crate instead of the `sr-sys` crate?

Reply:
Don't see why not, but why would we want to expose it? Seems to exist at the same level of abstraction as the AbiSchemaRegistryClient and will be accessed transparently through SchemaRegistryClient.

Reply:
Well that's the point - moving it into the sr crate means that it can be private. Right now it's public outside of the sr-system crate

Reply:
Ah, duh. The Abi impl is in here (and so public) to hide the actual ABI surface. Makes sense.

Reply:
done